### PR TITLE
Add default resource for StaticFiles

### LINKED ISSUE #57: StaticFiles handler will not invoke default_resource on 404
When using `fs::StaticFiles::new(".", false)` default resource will not be called upon hitting 404.

Presumably because we only in handler for non-default resource

### PR DESCRIPTION
**Pull request by DoumanAsh**
I'd like to hear opinion on my approach to default resource.

Currently tests are not working since it expects HttpResponse (and which also provides extra information about result)

Maybe you could advise test strategy for just `Reply`?

Closes #57

### CODE REVIEW

**Comment by fafhrd91** [comment]
I think approach is fine. 

I am not sure about `Resource`. I think it is too flexible.

**Comment by DoumanAsh** [comment]
Is there a proper way to test resulting `Reply` though?

For `Resource` though I only aimed for consistency with `Application::default_resource`

**Comment by fafhrd91** [comment]
1. for `Reply` you can convert it to a response and then inspect it or you can use TestServer.

2. `Resource` can lead to confusion, you can register route with wrong predicates. It is better to use Handler

**Comment by DoumanAsh** [comment]
Rebase, feel free to take a look

**Comment by fafhrd91** [comment]
I think it looks good. One question, should we make StaticFiles generic over application state?

**Comment by DoumanAsh** [comment]
I was thinking to avoid complicating default handler.
Do you think we need?
I can add if you'd like

**Comment by fafhrd91** [comment]
Let’s add it, just for sake of symmetry

**Review by fafhrd91 — APPROVED** [review]

**Comment by fafhrd91** [comment]
Could you also add entry to changes

**Comment by DoumanAsh** [comment]
As per your request: updated changes and added State to Static Files

File: src/fs.rs [hunk 1]
```diff
@@ -447,28 +428,32 @@ impl StaticFiles {
...
     /// showing files listing.
-    pub fn index_file<T: Into<String>>(mut self, index: T) -> StaticFiles {
+    pub fn index_file<T: Into<String>>(mut self, index: T) -> StaticFiles<S> {
         self.index = Some(index.into());
         self
     }
+
+    /// Sets default resource which is used when no matched file could be found.
+    ///
+    /// Contrary to Application, it cannot have state.
```
Comment (fafhrd91) [inline src/fs.rs]: remove this

**Comment by fafhrd91** [comment]
lgtm, just fix comment for default_handler

you should have write permission

**Pull request merged**